<!DOCTYPE html>
<html lang="zh-CN" data-bs-theme="dark">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link
      rel="icon"
      href="https://i.328888.xyz/2023/04/14/ixDL13.jpeg"
      type="image/png" />
    <script src="https://cdn.bootcdn.net/ajax/libs/jquery/3.6.4/jquery.min.js"></script>
    <link
      href="https://cdn.bootcdn.net/ajax/libs/twitter-bootstrap/5.3.0-alpha3/css/bootstrap.css"
      rel="stylesheet" />
    <script src="https://cdn.bootcdn.net/ajax/libs/twitter-bootstrap/5.3.0-alpha3/js/bootstrap.bundle.js"></script>
    <title>
      8-6promise对象的静态构造方法Promise.all()所有promise都成功才返回结果,否则返回第一个据因
    </title>
  </head>
  <body>
    <p>
      Promise.all()方法是一个JavaScript的全局对象，它可以接收<strong>一个promise的可迭代类型的输入</strong>，比如数组，Map，Set等，然后返回一个Promise实例，这个实例的结果也是一个数组，包含了输入数组中的所有promise的结果
    </p>
    <ol>
      <li>
        如果输入的可迭代对象为空，Promise.all()会同步地返回一个已完成（resolved）状态的promise。
      </li>
      <li>
        如果输入的可迭代对象不包含任何promise，或者所有输入的promise都完成（resolved），Promise.all()会异步地返回一个已完成（resolved）状态的promise。
      </li>
      <li>
        如果输入的可迭代对象中有任何一个promise失败（rejected），Promise.all()会异步地返回一个失败（rejected）状态的promise，并且拒绝（reject）的拒因是第一个失败的promise的原因。
      </li>
    </ol>
    <ol>
      Promise.all()方法有一些应用场景，比如：
      <li>
        多个请求成功的结果合并在一起。如果一个页面有多个请求，我们需要所有的请求都成功返回数据后再一起处理渲染,但凡出了错.就直接进入catch()
      </li>
      <li>
        合并请求结果并处理错误。如果我们需要单独处理一个请求的数据渲染和错误处理逻辑，有多个请求，我们就需要在多个地方写
      </li>
      <li>
        验证多个请求结果是否都是满足条件。如果我们需要对表单的输入内容进行安全验证，调用的是云函数写的方法，表单有多个字段需要验证，都是调用的一个内容安全校验接口，全部验证通过则可以进行正常的提交
      </li>
    </ol>
    <script>
      let p1 = new Promise((resolve, reject) => {
        let num = Math.random();
        num > 0.5 ? resolve("1成功:" + num) : reject("1失败:" + num);
      });
      let p2 = new Promise((resolve, reject) => {
        let num = Math.random();
        num > 0.5 ? resolve("2成功:" + num) : reject("2失败:" + num);
      });
      let p3 = new Promise((resolve, reject) => {
        let num = Math.random();
        num > 0.5 ? resolve("3成功:" + num) : reject("3失败:" + num);
      });
      console.log("p1=>", p1);
      console.log("p2=>", p2);
      console.log("p3=>", p3);
      Promise.all([p1, p2, p3])
        .then((res) => {
          console.log("res=>", res);
        })
        .catch((err) => {
          console.log("err=>", err);
        });
    </script>
  </body>
</html>
